Jhancock/text UI tests 2

JonathanHancock0 opened the pull request:
Created a Terminal User Interface for nanorc, which runs in the terminal when using the --tui option. Information displayed to the user includes the run number and status (running or not running), as well as the current FSM state and a tree view of applications. Buttons are dynamically generated for all currently allowed commands, allowing the user to execute them with a click. An additional dialogue can be brought up for entering parameters, and always appears when one or more parameters are mandatory. Logs from nanorc are also displayed, and can be saved to a file in the CWD.

JonathanHancock0 committed:
Initial commit part 1

JonathanHancock0 committed:
Initial commit part 2

JonathanHancock0 committed:
The user can toggle whether or not to always bring up the input dialogue. Commands are disabled while nanorc is working.

JonathanHancock0 committed:
Added a tree display for the DAQ apps

JonathanHancock0 committed:
TUI now displays logs from the sending of commands, albeit with no scrolling

JonathanHancock0 committed:
Altered the API so that information about node state is transferred, and added this info (and colours) to the tree display

JonathanHancock0 committed:
Fixed the sending of logs from the input window, and the handling of mandatory & not provided arguments.

JonathanHancock0 committed:
Changes to layout, and better value handling in the rest code

JonathanHancock0 committed:
Changed how the TUI gets the list of allowed commands, it can now access them all.\nBug fixes relating to argument parsing in the API.

JonathanHancock0 committed:
Run info display is now fully functional.
Colour indicates whether the RC is running or not, and run number and type is provided if so.
Float validation changed to use try/except, as the old method would accept things like 1.1.1.1.1 as inputs.

JonathanHancock0 committed:
Made command HTTP requests asynchronous with HTTPX, in order to stop freezing.\nAlso refactored how the information about whether the system is idle or not is propagated, so that we can force this to take place before the command is sent.

JonathanHancock0 committed:
Fixed ordering of logs

JonathanHancock0 committed:
Cleanup of unneeded code

JonathanHancock0 committed:
Fixed CSS so that the second row of buttons displays.
Re-added the quit button (calls shutdown, and then closes the TUI even if it fails).
Fixed a bug where the cancel button would crash the TUI.
Exceptions in the REST now appear in the input window error display, instead of causing a ValueError.

JonathanHancock0 committed:
Made most requests asynchronous, realigned bottom row of widgets.

JonathanHancock0 committed:
Minor layout changes

JonathanHancock0 committed:
Removed todo list from the bottom

JonathanHancock0 committed:
Fixed display issues with logs

JonathanHancock0 committed:
Tree view now immediately displays if an app dies

plasorak commented on src/nanorc/tui.py:
```diff
@@ -0,0 +1,589 @@
+        time = time[:-7]               # Times to the nearest second instead of microsecond
+        time = "-".join(time.split())  # Joins date and time with a hyphen instead of a space
+        time = time.replace(":","")    # chuck the weird ":"
+        filename = f"logs_{time}"
```
Could we make this a `.txt` file?

plasorak commented on src/nanorc/tui.py:
```diff
@@ -0,0 +1,589 @@
+    def save_logs(self) -> None:
+        data = self.logs
+        time = str(datetime.now())
+        time = time[:-7]               # Times to the nearest second instead of microsecond
```
There's probably a better way to do that and the line bbelow, with datetime

JonathanHancock0 replied:
Changed it and altered the formatting to YYYY-MM-DD-XXhYYmZZs (with the x y and z being hours minutes and seconds)

plasorak commented on src/nanorc/tui.py:
```diff
@@ -0,0 +1,589 @@
+
+    def update_ask(self):
+        if alwaysAsk:
+            self.ask = "All inputs required"
```
What is this one intented to show?

JonathanHancock0 replied:
The input can be in two modes: one where we always bring up the input windown, and one where we only bring it up when there's mandatory arguments. This function shows a string corresponding to the current mode in the status box.

plasorak replied:
Ok, so basically the popup with arguments will only appear if the TUI is in "All Input Required" (unless there are mandatory arguments)? I think that should be labelled "expert mode", and "non-expert mode".

JonathanHancock0 replied:
Ok, changed it.

plasorak commented on src/nanorc/tui.py:
```diff
@@ -0,0 +1,589 @@
+class Command(Static):
+    commands = reactive([])
+
+    class NewLog(Message):    
```
Can this one go with the Logs class, rather than here?

JonathanHancock0 replied:
Message classes need to be defined in the body of the class that produces them, so NewLog has to be defined inside of Command, since the command widget is what is sending the message.

plasorak commented on src/nanorc/tui.py:
```diff
@@ -0,0 +1,589 @@
+
+    def watch_commands(self, commands:list[str]) -> None:
+        second_line = ["exclude", "include", "pin_threads"]
+        box1 = self.query(Horizontal)[0]
```
This one and the line after are a bit scary, they rely on the ordering of the DOM. Could you change that to have a more specific query? There are plenty of queries available in here: https://textual.textualize.io/guide/queries/

JonathanHancock0 replied:
Used a query_one that asks for an id instead

plasorak commented on src/nanorc/tui.py:
```diff
@@ -0,0 +1,589 @@
+                try:                                    #Try to shutdown
+                    payload = {'command': button_id}
+                    async with httpx.AsyncClient() as client:
+                        r2 = await client.post(f'{self.hostname}/nanorcrest/command', auth=("fooUsr", "barPass"), data=payload, timeout=60)
```
Maybe it would be good to have this authentication centralised

JonathanHancock0 replied:
Moved the tuple into a global variable.

plasorak commented on src/nanorc/tui.py:
```diff
@@ -0,0 +1,589 @@
+        self.active = new_active
+                   
+class InputWindow(Widget):
+    class NewLog(Message):    
```
Do you need to redefine this one?

JonathanHancock0 replied:
Annoyingly it seems like that is the case: both the command widget and the input window produce this message so it needs to be defined in both, even though the functionality is identical.

plasorak commented on src/nanorc/tui.py:
```diff
@@ -0,0 +1,589 @@
+            if 'timeout' in payload:
+                t = int(payload['timeout'])
+            else:
+                t = 180
```
Anyway we could use the timeout provided in the command line arguments here?

JonathanHancock0 replied:
HTTP timeout needs to be longer than nanorc timeout, as if both time out at the same time, we will either get a handleable error from the API, or the TUI will crash. Thus, we get the default timeout given to us the app by the TUI, and use that + 5s as our default. If a timeout argument is provided then we add 5 to that too. There is however an issue with boot: it is taking longer than its timeout value would imply! This probably implies some deeply buried issue in nanorc, but for now we will just cheat and add a minute instead when we are booting.

plasorak commented on src/nanorc/tui.py:
```diff
@@ -0,0 +1,589 @@
+            self.remove()
+        
+class NanoRCTUI(App):
+    CSS_PATH = __file__[:-2] + "css"      #Gets the full path for a file called tui.css in the same folder as tui.py
```
Ah... We should find a better way to do that with `importlibs`

JonathanHancock0 replied:
Pierre has done this with his new commit.

plasorak commented on src/nanorc/tui.py:
```diff
@@ -0,0 +1,589 @@
+                        return                              #An attempt at sending a command is over
+                else:
+                    #We enforce here that the string will be convertable to the appropriate type
+                    match self.params[i.id]['type']:
```
Can you move that validation code to a different function/file?

JonathanHancock0 replied:
Code now runs out of a seperate method of the inputwindow class

plasorak commented on src/nanorc/tui.py:
```diff
@@ -0,0 +1,589 @@
+            async with httpx.AsyncClient() as client:
+                r = await client.post(f'{self.hostname}/nanorcrest/command', auth=("fooUsr", "barPass"), data=payload, timeout=t)
+
+            if "logs" in r.json():
```
Can you parse the json only once? that should speed things up

JonathanHancock0 replied:
I've put the output of r.json() into a variable and used that instead of potentially converting from string to dict over and over. (this was done for other parts of the code too)

plasorak reviewed (commented):
A couple of trivial comments above, otherwise this looks very good

plasorak committed:
add the tui.css from importlib

JonathanHancock0 committed:
Made the changes suggested in review. The change in cli.py is to pass the default timeout to the TUI at startu.

JonathanHancock0 committed:
Made the mimimum allowed timeout 10 seconds.

plasorak committed:
a small adjustment to the run info

plasorak committed:
a bit of renaming

plasorak committed:
only display required inputs in non expert mode and make them red

JonathanHancock0 committed:
Added the --tui option to nano04rc and nanotimingrc, allowing them to be controlled with the TUI as well.
Modified the timeout behaviour for boot: for this command only, the read timeout is disabled. The others (connect, write and pool) use the same value as other commands.

JonathanHancock0 committed:
Merge branch 'jhancock/text-ui-tests-2' of github.com:DUNE-DAQ/nanorc into jhancock/text-ui-tests-2

JonathanHancock0 committed:
Added a TUI screenshot

JonathanHancock0 committed:
Update README.md

Added information about the TUI to the readme

JonathanHancock0 committed:
Added the "Choice" as a type in the input validation code. (used only for run type, which must be TEST or PROD)

JonathanHancock0 committed:
Merge branch 'jhancock/text-ui-tests-2' of github.com:DUNE-DAQ/nanorc into jhancock/text-ui-tests-2

Merging the documentation commits

JonathanHancock0 committed:
Fixed an incorrect variable in the last commit.

JonathanHancock0 committed:
Fixed another incorrectly named variable

plasorak committed:
more variant

plasorak committed:
add warning for delet logs and make quit primary

plasorak committed:
Merge branch 'develop' into jhancock/text-ui-tests-2

plasorak committed:
major bump

plasorak commented:
Unfortunately, it seems I cannot boot application now that I have merged develop in this branch, @JonathanHancock0 would you mind to have a look?

plasorak committed:
introduce hook to select when to kill application based on the parent thread, that gets set properly when running with the rest
